fs/cifs/nfs41acl_xdr.h

```diff
@@ -0,0 +1,125 @@
+#ifndef _NFS41ACL_H
+#define _NFS41ACL_H
+
+/*
+ * Native ZFS NFSv41-style ACL is packed in xattr as
+ * follows:
+ *
+ * struct nfsace4i {
+ *    uint32_t type; RFC 5661 Section 6.2.1.1
+ *    uint32_t flag; RFC 5661 Section 6.2.1.4
+ *    uint32_t iflag;
+ *    uint32_t access_mask; RFC 5661 Section 6.2.1.3
+ *    uint32_t who_id;
+ * };
+ *
+ * struct nfsacl4 {
+ *     uint32_t acl_flags; RFC 5661 Section 6.4.3.2
+ *     uint32_t ace_count;
+ *     struct nfsace4i aces<>;
+ * };
+ *
+ * iflag and who_id combined are sufficent for NFS server to convert into ACE
+ * who (RFC 5661 Section 6.2.1.5).
+ */
+
+#define NA41_NAME "system.nfs4_acl_xdr"
+#define NA_TYPE_OFFSET 0
+#define NA_FLAG_OFFSET 1
+#define NA_IFLAG_OFFSET 2
+#define NA_ACCESS_MASK_OFFSET 3
+#define NA_WHO_OFFSET 4
+
+#define ACE4_READ_DATA 0x00000001
+#define ACE4_WRITE_DATA 0x00000002
+#define ACE4_APPEND_DATA 0x00000004
+#define ACE4_READ_NAMED_ATTRS 0x00000008
+#define ACE4_WRITE_NAMED_ATTRS 0x00000010
+#define ACE4_EXECUTE 0x00000020
+#define ACE4_DELETE_CHILD 0x00000040
+#define ACE4_READ_ATTRIBUTES 0x00000080
+#define ACE4_WRITE_ATTRIBUTES 0x00000100
+#define ACE4_DELETE 0x00010000
+#define ACE4_READ_ACL 0x00020000
+#define ACE4_WRITE_ACL 0x00040000
+#define ACE4_WRITE_OWNER 0x00080000
+#define ACE4_SYNCHRONIZE 0x00100000
+
+#define ACE4_READ_PERMS  (ACE4_READ_DATA|ACE4_READ_ACL|ACE4_READ_ATTRIBUTES| \
+    ACE4_READ_NAMED_ATTRS)
+
+#define ACE4_WRITE_PERMS (ACE4_WRITE_DATA|ACE4_APPEND_DATA|ACE4_WRITE_ATTRIBUTES| \
+    ACE4_WRITE_NAMED_ATTRS)
+
+#define ACE4_MODIFY_PERMS (ACE4_READ_PERMS|ACE4_WRITE_PERMS|ACE4_SYNCHRONIZE| \
+    ACE4_EXECUTE|ACE4_DELETE_CHILD|ACE4_DELETE)
+
+#define ACE4_ALL_PERMS (ACE4_MODIFY_PERMS|ACE4_WRITE_ACL|ACE4_WRITE_OWNER)
+
+#define ACE4_FILE_INHERIT_ACE 0x00000001
```

Comment:
Should there be a comment here referring to RFC 5661 [6.2.1.4](https://datatracker.ietf.org/doc/html/rfc5661#section-6.2.1.4).  ACE flag